presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -121,16 +123,35 @@ public boolean isEnabled(Session session)
     @Override
     public Result apply(JoinNode joinNode, Captures captures, Context context)
     {
-        MultiJoinNode multiJoinNode = toMultiJoinNode(joinNode, context.getLookup(), getMaxReorderedJoins(context.getSession()));
+        // try reorder joins with projection pushdown first
+        MultiJoinNode multiJoinNode = MultiJoinNode.toMultiJoinNode(joinNode, context, true);
+        JoinEnumerationResult resultWithProjectionPushdown = chooseJoinOrder(multiJoinNode, context);
+        if (!resultWithProjectionPushdown.getPlanNode().isPresent()) {
+            return Result.empty();
+        }
+
+        if (!multiJoinNode.isPushedProjectionThroughJoin()) {
+            return Result.ofPlanNode(resultWithProjectionPushdown.getPlanNode().get());
+        }
+
+        // try reorder joins without projection pushdown
+        multiJoinNode = toMultiJoinNode(joinNode, context, false);
+        JoinEnumerationResult resultWithoutProjectionPushdown = chooseJoinOrder(multiJoinNode, context);
+        if (!resultWithoutProjectionPushdown.getPlanNode().isPresent()
```

Comment:
`!resultWithoutProjectionPushdown.getPlanNode().isPresent()` state seems not possible, isn't it? Maybe use `checkState` for this?